### PR TITLE
add development instructions

### PR DESCRIPTION
**Pull request by rvhonorato**
This PR adds instructions on how to develop haddock3 and also updates the `setup.py`  to include a `CustomDevelop` instruction class

### CODE REVIEW

**Review by VGPReys — DISMISSED** [review]

**Comment by rvhonorato** [comment]
After irl discussion it was decided we need to also add a "Developing HADDOCK3 in a HPC environment" section

**Comment by rvhonorato** [comment]
This will change after #1013 - best to solve that one first

File: DEVELOPMENT.md [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+#### OpenMPI
+
+On a **Linux-based** system:
+
+```bash
+sudo apt-get install openmpi-bin libopenmpi3 libopenmpi-dev
+```
+
+**Note** that this step is not required on OSX.
```
Comment (rvhonorato) [inline DEVELOPMENT.md:76]: Can someone with an OSX please confirm this? Sounds strange that openmi comes pre-installed @haddocking/haddock-developers

**Comment by amjjbonvin** [comment]
In my case the `pip install mpi4yp` did the trick (meaning by that the test are passing with `pytest`)

Testing MPI execution on a single machine does not make sense anyway as there is no guarantee it will work on multiple nodes.

But best if someone else confirms it.

**Comment by rvhonorato** [comment]
> Testing MPI execution on a single machine does not make sense anyway as there is no guarantee it will work on multiple nodes.

We do not test this, since we are not developing the mpi library. What it's tested is if our interface with the mpi software works - so having it installed is sufficient.

File: DEVELOPMENT.md (hunk 1 above)
Comment (VGPReys) [reply DEVELOPMENT.md:76]: Not functional.
I got an error running the `pip install mpi4py==3.1.6`
But `brew install mpi4py` was functional

Comment (VGPReys) [reply DEVELOPMENT.md:76]: But indeed, no need of openmpi things

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+    pytest-mock==3.12.0 \
+    fastapi==0.110.1 \
+    httpx==0.27.0 \
+    mpi4py==3.1.6
+```
+
+Install haddock3 in development mode.
+
+```bash
+python setup.py develop
```
Comment (VGPReys) [inline DEVELOPMENT.md:122]:
```bash
building 'haddock.bin.fast_rmsdmatrix' extension
clang -Wno-unused-result -Wsign-compare -Wunreachable-code -fno-common -dynamic -DNDEBUG -g -fwrapv -O3 -Wall -iwithsysroot/System/Library/Frameworks/System.framework/PrivateHeaders -iwithsysroot/Applications/Xcode.app/Contents/Developer/Library/Frameworks/Python3.framework/Versions/3.9/Headers -arch arm64 -arch x86_64 -Werror=implicit-function-declaration -Wno-error=unreachable-code -I/Users/vreys/Desktop/HADDOCK3/.venv/include -I/Library/Developer/CommandLineTools/Library/Frameworks/Python3.framework/Versions/3.9/include/python3.9 -c src/haddock/deps/fast-rmsdmatrix.c -o build/temp.macosx-10.9-universal2-3.9/src/haddock/deps/fast-rmsdmatrix.o -Wall -O3 -march=native -std=c99
error: unknown target CPU 'apple-m1'
note: valid target CPU values are: nocona, core2, penryn, bonnell, atom, silvermont, slm, goldmont, goldmont-plus, tremont, nehalem, corei7, westmere, sandybridge, corei7-avx, ivybridge, core-avx-i, haswell, core-avx2, broadwell, skylake, skylake-avx512, skx, cascadelake, cooperlake, cannonlake, icelake-client, rocketlake, icelake-server, tigerlake, sapphirerapids, alderlake, knl, knm, k8, athlon64, athlon-fx, opteron, k8-sse3, athlon64-sse3, opteron-sse3, amdfam10, barcelona, btver1, btver2, bdver1, bdver2, bdver3, bdver4, znver1, znver2, znver3, x86-64, x86-64-v2, x86-64-v3, x86-64-v4
error: command '/usr/bin/clang' failed with exit code 1
```

**Comment by amjjbonvin** [comment]
Ok - then we should add the recommendation to use brew to install mpi4py

**Comment by amjjbonvin** [comment]
I can’t reproduce this error.  May-be linked to the installed compilers and such?

Do you have xcode installed on your mac?
If not try installing it first. If this solves the problem we should make a note of it

File: DEVELOPMENT.md [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    pytest==7.3.1 \
+    pytest-cov==4.0.0 \
+    hypothesis==6.75.1 \
+    pytest-mock==3.12.0 \
+    fastapi==0.110.1 \
+    httpx==0.27.0 \
+    mpi4py==3.1.6
+```
+
+**Note**: On OSX, if the installation of mpi4py fails, use instead `brew install mpi4py`
```
Comment (VGPReys) [inline DEVELOPMENT.md:113]:
```suggestion
**Note**: On OSX, if the installation of mpi4py fails, run first `brew install mpi4py`
```

**Comment by amjjbonvin** [comment]
And another issue with the instructions is that we are missing a step about CNS - without the executable the tests will fail.

**Comment by VGPReys** [comment]
Also missing instructions for the softlink to CNS ?
```bash
ln -s /path/to/cns-os-cpu.exe src/haddock/bin/cns
```

**Comment by rvhonorato** [comment]
> Getting errors in the tests: `FAILED tests/test_cli_restraints.py::test_restrain_bodies - AssertionError: assert 'assign (segi...1.582 0.0 0.0' == 'assign (segi...6.542 0.0 0.0' FAILED tests/test_cli_restraints.py::test_restrain_bodies_exclude - AssertionError: assert 'assign (segi...7.660 0.0 0.0' == 'assign (segi...5.187 0.0 0`
> 
> As the residues are randomly selected, could this be the problem? E.g. a different pair is selected because of different random numbers or related?

Unclear to me how the different installations would fail this test.

**Comment by rvhonorato** [comment]
> Also missing instructions for the softlink to CNS ?
> 
> ```shell
> ln -s /path/to/cns-os-cpu.exe src/haddock/bin/cns
> ```

Why have a link and not use the pre-compiled ones?

**Comment by rvhonorato** [comment]
@haddocking/haddock-developers I have updated the instructions, please try again - make sure all your system dependencies are installed.

**Comment by rvhonorato** [comment]
@haddocking/haddock-developers I have updated the instructions, please try again - make sure all your system dependencies are installed.

**Review by amjjbonvin — APPROVED** [review]
Now working as a charm!

**Pull request merged**